### PR TITLE
TICK-413 Adds new models to tick for supervised learning

### PR DESCRIPTION
**Pull request by stephanegaiffas**
This adds a bunch of models in tick.optim.model

For binary classification :
* ModelHinge
* ModelQuadraticHinge
* ModelSmoothedHinge
* ModelModifiedHuber

For regression :
* ModelHuber
* ModelEpsilonInsensitive
* ModelAbsoluteRegression

It also cleans a lot the doctrines of the model classes, and gives more details in the tick.optim.model documentation

### CODE REVIEW

**Comment by stephanegaiffas** [comment]
@Mbompr @MaryanMorel @Dekken : this PR was some kind of a small marathon :)
It adds a bunch of linear models in tick.optim.model (7 to be precise, actually the most used ones), and adds a lot of details in sphinx doc and python docstrings (hence the 2500 lines, so don't be scared :))
Enjoy !

**Comment by Mbompr** [comment]
I have rebased this on #89 to fix the file not found problem. We need to wait for #89 to be merged before merging this one

**Comment by stephanegaiffas** [comment]
@Mbompr @MaryanMorel anyone ? :)

File: tick/optim/model/huber.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    for :math:`y, y' \\in \\mathbb R`, where :math:`\\delta > 0` can be tuned
+    using the ``threshold`` argument. Data is passed to this model through the
+    ``fit(X, y)`` method where X is the features matrix (dense or sparse) and
+    y is the vector of labels.
+
+    Parameters
+    ----------
+    fit_intercept : `bool`
+        If `True`, the model uses an intercept
+
```
Comment (MaryanMorel) [inline tick/optim/model/huber.py]: Missing doc for threshold

File: tick/optim/model/poisreg.py [hunk 2]
```diff
@@ -17,9 +17,34 @@
 class ModelPoisReg(ModelGeneralizedLinear,
                    ModelSecondOrder,
                    ModelSelfConcordant):
-    """Poisson regression model with identity or exponential link.
-    This class gives first order and second order information for this
-    model (gradient, loss and hessian norm).
+    """Poisson regression model with identity or exponential link for data with
+    a count label. This class gives first order and second order information for
+    this model (gradient, loss and hessian norm)  and can be passed
+    to any solver through the solver's ``set_model`` method.
```
Comment (MaryanMorel) [inline tick/optim/model/poisreg.py]: Maybe add a comment on overflows here?

File: tick/optim/model/huber.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        return self
+
+    def _grad(self, coeffs: np.ndarray, out: np.ndarray) -> None:
+        self._model.grad(coeffs, out)
+
+    def _loss(self, coeffs: np.ndarray) -> float:
+        return self._model.loss(coeffs)
+
+    def _get_lip_best(self):
+        # TODO: Use sklearn.decomposition.TruncatedSVD instead?
```
Comment (MaryanMorel) [inline tick/optim/model/huber.py]: Remaining TODO. What would be the advantages of doing so? Imo, relying on sklearn outside the inference submodule is not great, but I don't know what you decided on this matter. @Mbompr ?

File: doc/modules/optim.rst [hunk 4]
```diff
@@ -161,33 +161,228 @@ Model classes can be used with any solver class, by simply passing them using th
...
-.. _optim-model-glm:
+.. _linear-models:
 
-1.2. Generalized linear models
-------------------------------
+1.2. Linear models
+------------------
 
-We describe here generalized linear models for supervised learning.
+We describe here (generalized) linear methods for supervised learning.
```
[the remark this answers is not in the document]
Comment (stephanegaiffas) [reply doc/modules/optim.rst]: Well, because these are not really "models", for instance there is not model associated to the hinge loss... A big refactoring of model names and organization should be done maybe in another pull request (since involves refactoring and moving other models)

File: doc/modules/optim.rst [hunk 5]
```diff
@@ -161,33 +161,228 @@ Model classes can be used with any solver class, by simply passing them using th
...
 Given training data :math:`(x_i, y_i) \in \mathbb R^d \times \mathbb R`
-for :math:`i=1, \ldots, n`, we consider models with a goodness-of-fit that
-writes
+for :math:`i=1, \ldots, n`, we consider goodness-of-fit that writes
 
 .. math::
 	f(w, b) = \frac 1n \sum_{i=1}^n \ell(y_i, b + x_i^\top w),
 
-where :math:`w \in \mathbb R^d` is a vector containing the model weights,
+where :math:`w \in \mathbb R^d` is a vector containing the model-weights,
```
[the remark this answers is not in the document]
Comment (stephanegaiffas) [reply doc/modules/optim.rst]: dunna

File: tick/optim/model/absolute_regression.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    n_features : `int` (read-only)
+        Number of features
+
+    n_coeffs : `int` (read-only)
+        Total number of coefficients of the model
+
+    n_threads : `int`, default=1 (read-only)
+        Number of threads used for parallel computation.
+
+        * if ``int <= 0``: the number of physical cores available on
```
[the remark this answers is not in the document]
Comment (stephanegaiffas) [reply tick/optim/model/absolute_regression.py]: Indeed

File: tick/optim/model/absolute_regression.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+          the CPU
+        * otherwise the desired number of threads
+    """
+    def __init__(self, fit_intercept: bool = True, n_threads: int = 1):
+        ModelFirstOrder.__init__(self)
+        ModelGeneralizedLinear.__init__(self, fit_intercept)
+        self.n_threads = n_threads
+
+    # TODO: implement _set_data and not fit
```
[the remark this answers is not in the document]
Comment (stephanegaiffas) [reply tick/optim/model/absolute_regression.py]: This will remain so: to be done in another PR

File: tick/optim/model/huber.py (hunk 1 above)
Comment (stephanegaiffas) [reply tick/optim/model/huber.py]: Done

File: tick/optim/model/huber.py (hunk 3 above)
Comment (stephanegaiffas) [reply tick/optim/model/huber.py]: Same as before, to be decided in another PR. Let's keep it like this for now (was already there before for other models...)

File: tick/optim/model/poisreg.py (hunk 2 above)
Comment (stephanegaiffas) [reply tick/optim/model/poisreg.py]: Done
